### PR TITLE
fix(grpc): resolve ambiguous types

### PR DESCRIPTION
**Pull request by ssddOnTop**
## Summary by CodeRabbit

- **New Features**
	- Enhanced configuration handling with methods to resolve ambiguous types and improve connections between different data types.

### CODE REVIEW

File: src/config/config_module.rs [hunk 1]
```diff
@@ -173,6 +173,45 @@ fn get_output_types(config: &Config, input_types: &HashSet<String>) -> HashSet<S
     types
 }
 
+pub struct Resolution {
+    pub input: String,
+    pub output: String,
+}
+
+impl ConfigModule {
+    pub fn resolve_ambiguous_types(mut self, resolver: impl Fn(&str) -> Resolution) -> Self {
```
Comment (tusharmath) [inline src/config/config_module.rs:182]: add a doc.

Comment (ssddOnTop) [reply src/config/config_module.rs:182]: @coderabbitai generate documentation

**Comment by ssddOnTop** [comment]
@coderabbitai generate tests

File: src/config/config_module.rs [hunk 2]
```diff
@@ -115,68 +114,57 @@ impl Deref for ConfigModule {
...
-    for (type_name, type_of) in config.types.iter() {
-        if (type_of.interface || !type_of.fields.is_empty()) && !input_types.contains(type_name) {
-            for (_, field) in type_of.fields.iter() {
-                types.insert(field.type_of.clone());
+                    for arg in field.args.values_mut() {
+                        if arg.type_of.eq(current_name) {
+                            arg.type_of = input_name.clone();
+                        }
+                    }
+                }
```
Comment (tusharmath) [inline src/config/config_module.rs:156]: We need to recursively update the type name wherever referenced.

File: src/config/config.rs [hunk 3]
```diff
@@ -727,6 +725,70 @@
...
+            if (type_of.interface || !type_of.fields.is_empty()) && !input_types.contains(type_name)
+            {
+                for (_, field) in type_of.fields.iter() {
+                    types.insert(field.type_of.clone());
+                }
+            }
+        }
+
+        types
+    }
```
[the remark this answers is not in the document]
Comment (ssddOnTop) [reply src/config/config.rs:781]: give more tests

File: src/config/config_module.rs [hunk 4]
```diff
@@ -115,69 +114,202 @@
...
+        };
+
+        config_module = config_module.resolve_ambiguous_types(resolver);
+
+        assert!(config_module.config.types.contains_key("Type1Input"));
+        assert!(config_module.config.types.contains_key("Type1Output"));
+        assert!(config_module.config.types.contains_key("Type2Output"));
+        assert!(config_module.config.types.contains_key("Type2Input"));
+        assert!(config_module.config.types.contains_key("Type3"));
+    }
```
[the remark this answers is not in the document]
Comment (ssddOnTop) [reply src/config/config_module.rs:311]: add these tests

**Comment by shashitnak** [comment]
@ssddOnTop This is an example config which was generated for `news.proto` file. It doesn't handle the case where same type is used as input as well as output

`news.proto`
```proto
syntax = "proto3";

import "google/protobuf/empty.proto";

package news;

enum Status {
  PUBLISHED = 0;
  DRAFT = 1;
  DELETED = 2;
}

message News {
  int32 id = 1;
  string title = 2;
  string body = 3;
  string postImage = 4;
  Status status = 5;
}

service NewsService {
  rpc GetAllNews(google.protobuf.Empty) returns (NewsList) {}
  rpc GetNews(NewsId) returns (News) {}
  rpc GetMultipleNews(MultipleNewsId) returns (NewsList) {}
  rpc DeleteNews(NewsId) returns (google.protobuf.Empty) {}
  rpc EditNews(News) returns (News) {}
  rpc AddNews(News) returns (News) {}
}

message NewsId { int32 id = 1; }

message MultipleNewsId { repeated NewsId ids = 1; }

message NewsList { repeated News news = 1; }
```

generated `.graphql` file
```graphql
schema @server @upstream @link(src: "../rust-grpc/news.proto", type: Protobuf) {
  query: Query
}

input INPUT_News__News @tag(id: "news.News") {
  body: String
  id: Int
  postImage: String
  status: News__Status
  title: String
}

input News__MultipleNewsId @tag(id: "news.MultipleNewsId") {
  ids: [News__NewsId]!
}

input News__NewsId @tag(id: "news.NewsId") {
  id: Int
}

enum News__Status @tag(id: "news.Status") {
  DELETED
  DRAFT
  PUBLISHED
}

type News {
  addNews(news: INPUT_News__News!): News__News! @grpc(method: "news.NewsService.AddNews")
  deleteNews(newsId: News__NewsId!): Empty! @grpc(method: "news.NewsService.DeleteNews")
  editNews(news: INPUT_News__News!): News__News! @grpc(method: "news.NewsService.EditNews")
  getAllNews: News__NewsList! @grpc(method: "news.NewsService.GetAllNews")
  getMultipleNews(multipleNewsId: News__MultipleNewsId!): News__NewsList! @grpc(method: "news.NewsService.GetMultipleNews")
  getNews(newsId: News__NewsId!): News__News! @grpc(method: "news.NewsService.GetNews")
}

type News__NewsList @tag(id: "news.NewsList") {
  news: [News__News]!
}

type Query {
  news: News
}
```

**Review by tusharmath — APPROVED** [review]

**Review by tusharmath — APPROVED** [review]

**Pull request merged**